### PR TITLE
Keep shipped public API entries when recording removals

### LINKED ISSUE #4638: PublicAPI baseline sync orphans *REMOVED* markers by deleting shipped entries immediately
## Summary

`tools/ModularPipelines.OptionsGenerator/scripts/Merge-PublicApiBaselineSnapshot.ps1` (driven by `Sync-PublicApiBaselines.ps1`, introduced for #4447) records a removed public member by **both** deleting its line from `PublicAPI.Shipped.txt` **and** appending `*REMOVED*<symbol>` to `PublicAPI.Unshipped.txt` in the same commit. `Test-SyncPublicApiBaselines.ps1` / `Test-MergePublicApiBaselineSnapshot.ps1` encode that transition.

Under the `Microsoft.CodeAnalysis.PublicApiAnalyzers` (5.6.0) lifecycle that #4244 adopted ("baselined at the v4 cut ... a CI step should fail when `PublicAPI.Unshipped.txt` is non-empty on `main` after a release"), a `*REMOVED*` marker is meant to sit alongside the still-present `Shipped.txt` entry until a release "ships" the unshipped file and collapses the pair. Deleting the shipped line immediately leaves the marker orphaned from the moment it is committed: the analyzer treats it as inert (no RS0017/RS0024), but it can never be reconciled by tooling.

Observed on `main` today (raised by the automated review on #4558): `src/ModularPipelines.Kubernetes/PublicAPI.Unshipped.txt` carries 202 `*REMOVED*` lines and none of them still has a matching line in `PublicAPI.Shipped.txt`. Every automated refresh compounds this (44 more on #4558 alone), so it should be fixed in the script rather than accepted indefinitely. #4558 and the other automated refresh PRs are mechanically correct under the current script and should not hand-edit generated baselines (CLAUDE.md).

## Decision needed (pick one, then implement it in the script + tests)

- **(a) Follow the analyzer lifecycle (matches #4244):** on removal, leave the `Shipped.txt` entry in place and only add the `*REMOVED*` marker to `Unshipped.txt`; on signature change, add the new signature to `Unshipped.txt` and a `*REMOVED*` marker for the old one, again leaving `Shipped.txt` untouched. A future release step (or a one-off migration) reconciles the pair. Existing orphaned markers need a one-time cleanup or a matching re-insertion so the files become consistent.
- **(b) Treat `Shipped.txt` as a rolling current-surface snapshot:** keep deleting shipped lines but stop writing `*REMOVED*` markers, so `Unshipped.txt` only lists additions. Simpler, but abandons the ship/unshipped lifecycle that #4244 asked for and loses the removal changelog.

Option (a) is the one consistent with the documented analyzer contract and #4244; (b) only makes sense if the repo decides it will never run a ship step.

## Acceptance criteria

- `Merge-PublicApiBaselineSnapshot.ps1` implements the chosen convention; `Test-MergePublicApiBaselineSnapshot.ps1` and `Test-SyncPublicApiBaselines.ps1` assert it for removal, signature change, empty input and idempotency.
- A one-time migration (script or generated PR) brings the existing `PublicAPI.*.txt` files on `main` into the chosen shape, so no orphaned `*REMOVED*` markers remain.
- The generated-PR summary (#4576) still reports removals correctly under the new shape.
- Regenerated baselines from the Generate CLI Options workflow build clean (no RS0016/RS0017/RS0024) for at least one integration package.

Raised by the automated review on #4558; pre-existing since #4470 / #4447.

### PR DESCRIPTION
**Pull request by thomhurst**
Public API synchronization retains shipped declarations while recording their removal, restores shipped history from orphaned removal markers, and removes stale markers when an API is reintroduced. CI rejects orphaned markers, markers in shipped files, and duplicate entries. Release-time collapse of shipped/removal pairs remains tracked by #3997.

The shared parser now owns blank/header/entry classification as well as the removal prefix. Merge, history repair, validation, and summaries use the same rules. API line text remains exact, matching the pinned analyzer's removal recognition and payload comparison; leading or trailing whitespace is not normalized into a different API. Negative validator tests capture each child process's exit code while preserving the caller's previous code.

This branch is rebased onto main at b9e1a0f178. The previous Ubuntu failure was the command-ordering test waiting for a file logger to contain completion output. That test now records logger calls directly, retains its ordering assertions, requires successful command execution, and disposes the pipeline in a finally block.

Validation at e2bc359e1b9099d0af89eb0d5ffa705e6f44a7e2:
- Baseline validator passes for all 61 package projects.
- Six PowerShell suites pass: shared line/marker parsing, baseline validation, merge/reintroduction/idempotency, history repair, API summary, and synchronization orchestration.
- Regressions reproduced inconsistent handling of an indented header and leaking native exit-code state before the fixes. Existing exact-marker tests remain intact.
- Release core/test solution builds with 181 warnings and zero errors; all 52 command logger tests pass.
- Scoped formatting and git diff --check pass. The formatter reports its existing unsupported F# fixture project.
- All local dotnet validation retains the default 10-minute/2-GB guard.

Earlier authoritative jq children were incorporated, but this rebase and parser update change generator inputs. Final jq generation succeeded at https://github.com/thomhurst/ModularPipelines/actions/runs/34797964854 and its provenance-only refresh was incorporated through #4913 after all 11 checks passed and reviews cleared it. Current parent head: 0f06c406226fc8fd2fb1f60f7010eb162855ad7c. Fresh parent CI and review are required before merge. No generated baseline or provenance was hand-edited.

Closes #4638

### CODE REVIEW

**Comment by thomhurst** [comment]
Addressed both carried-over findings at exact head a68ce8d20a83f07577b99e7a5546f293886d211e.

**1. `*REMOVED*` convention triplicated.** Added `scripts/PublicApiRemovedMarker.ps1` exposing `Get-RemovedMarkerPrefix`, `Test-RemovedMarker`, `Get-RemovedMarkerEntry` and `New-RemovedMarker`, and dot-sourced it from `Assert-PublicApiBaselines.ps1`, `Merge-PublicApiBaselineSnapshot.ps1` and `Write-PublicApiChangeSummary.ps1`. None of the three defines the prefix or its parse rule any more. The helper is covered by `scripts/Test-PublicApiRemovedMarker.ps1` (prefix, round-trip, ordinal/no-trim rejection cases, guard on non-markers).

**2. Duplicated per-project path derivation in the assert script.** The missing-baseline and orphaned-marker checks are now one pass over `$packageProjects`: each project derives its two baseline paths once, a project with a missing file is recorded and skipped, and the throw order (missing first, then orphans) is unchanged so the existing test still discriminates both failures.

Also wired `Test-PublicApiRemovedMarker.ps1` and the pre-existing `Test-AssertPublicApiBaselines.ps1` (which CI never ran) into `dotnet.yml` ahead of the baseline coverage step.

Local runs: the five script tests (`Test-PublicApiRemovedMarker`, `Test-AssertPublicApiBaselines`, `Test-MergePublicApiBaselineSnapshot`, `Test-SyncPublicApiBaselines`, `Test-WritePublicApiChangeSummary`) all pass, and `Assert-PublicApiBaselines.ps1` verifies 61 package projects on this head.

@claude please re-review exact head a68ce8d20a83f07577b99e7a5546f293886d211e.

**Comment by thomhurst** [comment]
Correction and resync at exact head `42a24ec49e`. My earlier resync to `8fc873a441` was made from a stale local branch and dropped the `a68ce8d20a` refactor (shared `scripts/PublicApiRemovedMarker.ps1` helper, single-pass assert script, CI wiring for the two script tests). That commit is restored: the branch is now `fix(publicapi)` + `refactor(publicapi)` rebased onto current `main` (Argo CD baseline re-migrated after #4535), plus one more commit that adds `scripts/PublicApiRemovedMarker.ps1` to `Get-GeneratedOptionsSourcePath` and the Generate CLI Options push paths, as the Codex thread asked. Local: all six baseline/marker/provenance script tests pass and `Assert-PublicApiBaselines.ps1` verifies 61 packages. @claude please re-review exact head 42a24ec49ea956cb441ad497f32d010e607c6a2b.

File: tools/ModularPipelines.OptionsGenerator/scripts/Merge-PublicApiBaselineSnapshot.ps1 [hunk 1]
```diff
@@ -9,7 +9,7 @@ param(
 )
 
 $ErrorActionPreference = 'Stop'
-$removedPrefix = '*REMOVED*'
+. (Join-Path $PSScriptRoot '../../../scripts/PublicApiRemovedMarker.ps1')
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply tools/ModularPipelines.OptionsGenerator/scripts/Merge-PublicApiBaselineSnapshot.ps1:12]: Fixed in 42a24ec49e: scripts/PublicApiRemovedMarker.ps1 is now listed in Get-GeneratedOptionsSourcePath (scripts/GeneratedOptionsProvenance.ps1) and in the workflow's push.paths, so a helper-only change changes the generator fingerprint and triggers regeneration. Test-GeneratedOptionsProvenance.ps1 cross-checks both lists.

**Comment by thomhurst** [comment]
Addressed the remaining finding at exact head `beb88f6e8e` (rebased onto current `main` after #4462, conflict-free). The duplicated `*REMOVED*...DequeueModuleAsync` line in `src/ModularPipelines/PublicAPI.Unshipped.txt` predates this PR (it is present on `main` too), so the migration did not create it; it is now removed, and a `sort | uniq -d` sweep across every `PublicAPI.*.txt` finds no other repeats. To keep that true, `Assert-PublicApiBaselines.ps1` now also fails on any entry listed twice in a baseline (bulk set-count fast path, per-line report only on mismatch; the tree check still runs in ~2.6 s for 61 packages), with a case in `Test-AssertPublicApiBaselines.ps1`. All six script tests pass locally. @claude please re-review exact head beb88f6e8eb77499dfa20cfd552fccf1eef15fbb.

File: scripts/Assert-PublicApiBaselines.ps1 [hunk 2]
```diff
@@ -25,21 +26,77 @@ $packageProjects = @(
...
-        if (-not (Test-Path -LiteralPath $baselinePath -PathType Leaf)) {
-            [System.IO.Path]::GetRelativePath($repositoryRootPath, $baselinePath)
+    # Shipped baselines run to 100k+ lines, so only read them when a marker needs checking.
+    $shipped = [System.Collections.Generic.HashSet[string]]::new(
+        [System.IO.File]::ReadAllLines($shippedPath),
+        [System.StringComparer]::Ordinal)
+    $relativePath = [System.IO.Path]::GetRelativePath($repositoryRootPath, $unshippedPath)
+    foreach ($marker in $markers) {
+        if (-not $shipped.Contains((Get-RemovedMarkerEntry $marker).Trim())) {
+            $orphanedMarkers.Add("${relativePath}: $($marker.Trim())")
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply scripts/Assert-PublicApiBaselines.ps1]: Fixed in 58fc729885: the marker lookup now uses Get-RemovedMarkerEntry unmodified (no Trim), matching Merge-PublicApiBaselineSnapshot.ps1, and the duplicate check compares exact lines too. Test-AssertPublicApiBaselines.ps1 covers a marker with trailing whitespace being reported as an orphan.

File: scripts/Assert-PublicApiBaselines.ps1 [hunk 3]
```diff
@@ -25,21 +26,77 @@ $packageProjects = @(
...
+            $missingBaselines.Add([System.IO.Path]::GetRelativePath($repositoryRootPath, $path))
+        }
 
-    foreach ($fileName in 'PublicAPI.Shipped.txt', 'PublicAPI.Unshipped.txt') {
-        $baselinePath = Join-Path $projectDirectory $fileName
+        continue
+    }
+
+    Add-DuplicateEntries $shippedPath
+    Add-DuplicateEntries $unshippedPath
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply scripts/Assert-PublicApiBaselines.ps1]: Fixed in 58fc729885: the per-project pass now reports a plain entry that appears in both PublicAPI.Shipped.txt and PublicAPI.Unshipped.txt (RS0025), while a shipped entry plus its *REMOVED* marker stays valid. Covered by a new case in Test-AssertPublicApiBaselines.ps1.

**Comment by thomhurst** [comment]
Addressed both CodeRabbit findings at exact head `58fc729885`: markers now compare exactly against shipped entries (no trimming, matching the merge script), and a plain symbol present in both baselines is reported as an RS0025 duplicate while the shipped-entry-plus-marker pair remains valid. Both files are read once per project. `Test-AssertPublicApiBaselines.ps1` covers both cases; the tree check still passes for 61 packages in ~3.5 s. @claude please re-review exact head 58fc7298859e51307b66ab8459b37d8da10595fc.

**Comment by thomhurst** [comment]
Fixed the Fast fail failure at exact head `5dac5398f4` (also rebased onto current `main` after #4656, conflict-free). `Test-AssertPublicApiBaselines.ps1` matched the assert script's message against the console-rendered output, which PowerShell wraps at 80 columns with a gutter on Linux, so the long cross-file-duplicate message never matched there. The two long-message cases now run the script in-process and inspect the thrown message. No product change; all script tests and the 61-package tree check pass locally.

**Comment by thomhurst** [comment]
CI on 5dac5398f4 failed in the `Test public API baseline checks` step after both scripts printed success: the last negative case in `Test-AssertPublicApiBaselines.ps1` runs the gate in a child `pwsh` that is expected to fail, leaving `\0` at 1, which the Actions pwsh shell reports as the step's exit code. 89864c6119 ends the script with an explicit `exit 0` (emulated the wrapper locally: exit 0). It also picks up the review's whitespace observation: `Test-ApiEntry` now treats whitespace-only lines as blank like `Read-Baseline` in the merge script, with a test case. Real gate on this checkout: 61 package projects verified.

@claude please re-review exact head 89864c61193c91ff3526737f9b0c58bc14987d83.

**Comment by thomhurst** [comment]
Merged `main` (brew #4621 and pulumi #4637 regenerations conflicted on their `PublicAPI.Unshipped.txt`). Resolution takes main's regenerated files and applies the PR's one-time migration rule to them: markers whose shipped entry is already gone are dropped, so the new orphan check passes. `Assert-PublicApiBaselines.ps1` on the merged tree: 61 package projects verified; the baseline test step exits 0.

@claude please re-review exact head 1449139521fbd4952cd65aab231dd4daae7ceab2.

**Comment by thomhurst** [comment]
Took all three follow-ups in 592e246522: `Assert-PublicApiBaselines.ps1` now reports every populated category (missing baselines, orphaned markers, duplicates) in one thrown message; the cross-file check inspects each distinct unshipped entry once, so a repeated-and-shipped entry yields one intra-file plus one cross-file finding rather than one per occurrence; and the test script resets `LASTEXITCODE` instead of `exit 0` so later scripts in the same step still run. Two new negative cases cover the combined report and the de-duplicated finding count. Emulated step exit 0 with a trailing statement executing; real gate 61 projects.

@claude please re-review exact head 592e246522c843324db3954a18b20ca4689b7f72.

**Comment by thomhurst** [comment]
Took the remaining nice-to-have in b2867d485a: a `*REMOVED*` marker found in `PublicAPI.Shipped.txt` is now reported with the orphaned markers (`... (marker in PublicAPI.Shipped.txt)`), with a test-script case. Emulated step exit 0; real gate 61 projects.

@claude please re-review exact head b2867d485ae015f44d1098466cdb41191c8af9d1.

**Comment by thomhurst** [comment]
Merged `main` after #4609 landed (conflict in `Test-MergePublicApiBaselineSnapshot.ps1`: both sides added cases at the end; kept both, the no-confirmed-removals case first, then #4609's empty-input cases). All five baseline test scripts and the real gate pass locally.

@claude please re-review exact head 1c9464cf4a3bfa7150894934895b664db9eb8fce.

**Comment by thomhurst** [comment]
Merged `main` again (go #4686 regeneration conflicted on its `PublicAPI.Unshipped.txt`; resolved like brew/pulumi/flux: main's file minus markers whose shipped entry is gone). Gate 61 projects; baseline test step exit 0.

@claude please re-review exact head 1fc04bce8be4500cbd8cd2eae6d8c00ad0738a8e.

**Comment by thomhurst** [comment]
Rebased onto current main and pushed add22e0a55. Baseline conflicts use the current declarations with only orphaned/duplicate markers removed. Both appended test groups survive. The marker helper remains in generator provenance; the newer monthly/manual scheduling policy is preserved without restoring push triggers.

All seven focused PowerShell suites and the validator for 61 package projects pass. The guarded core Release build passes with zero warnings/errors, formatting completes without code changes, and git diff --check passes. The previously reviewed script implementations are unchanged by this rebase. Waiting for the new CI/review cycle before merging.

**Comment by thomhurst** [comment]
Synced main to include the CodeQL extraction diagnostics from #4853. The previous failure was shared issue #4800: internal restore failures followed by an empty C# database (run 34762723397, job 103738257675). The new run can retain the underlying extraction logs. All ten PR commits are patch-equivalent after rebase (git range-diff); no implementation changed. All seven focused PowerShell suites and validation of 61 package baselines pass again. Prior build/format validation still applies to the unchanged implementation. New head: 48a603f023.

**Comment by thomhurst** [comment]
The refreshed C# CodeQL check passes. The remaining Ubuntu failure is the unrelated DistributedMasterAddsOnlyMissingUnmatchedWorkerMetrics(3, 2, 6, 3) core test at its 2-second CompleteAsync wait: https://github.com/thomhurst/ModularPipelines/actions/runs/34763889509/job/103743553432. PR #4678 already fixes this exact guard. Deferring the CI repair until that prerequisite merges; no blind rerun or duplicate timeout patch.

File: tools/ModularPipelines.OptionsGenerator/scripts/Merge-PublicApiBaselineSnapshot.ps1 [hunk 4]
```diff
@@ -102,19 +101,20 @@ foreach ($entry in $currentApis) {
...
-    if (-not $entry.StartsWith($removedPrefix, [System.StringComparison]::Ordinal)) {
+    if (-not (Test-RemovedMarker $entry)) {
         continue
     }
 
-    $removedApi = $entry.Substring($removedPrefix.Length)
-    if (-not $currentApis.Contains($removedApi)) {
+    $removedApi = Get-RemovedMarkerEntry $entry
+    if ($shippedApis.Contains($removedApi) -and -not $currentApis.Contains($removedApi)) {
         $null = $unshippedApis.Add($entry)
```
[the remark this answers is not in the document]
Comment (thomhurst) [reply tools/ModularPipelines.OptionsGenerator/scripts/Merge-PublicApiBaselineSnapshot.ps1]: Fixed in 69b264203b01536e66e8873b46b999cb9bdf41a6. The merge script restores shipped entries from historical orphaned markers before classifying the current snapshot. The migration was rerun from current main using Repair-PublicApiBaselineHistory.ps1: 23,267 shipped entries restored, all 23,446 unique removal markers retained, and one duplicate marker removed. An independent audit verified all 61 baseline pairs retain exactly the same active APIs and pending history. Historical and reintroduced-API tests fail before the fix and pass afterward; Chocolatey builds without RS0016/RS0017/RS0024 and all six integration tests pass.

**Comment by thomhurst** [comment]
Addressed the summary parsing finding in review 5191040103 with 69b264203b: marker detection and payload extraction now use the raw line. A trailing-space marker regression failed before the fix and passes afterward.

The requested generated-integration allowlist addition would weaken validation. `Resolve-GeneratedIntegrationValidation` returning `IsGeneratedIntegration=false` selects the full pipeline; it does not reject the PR. I verified that a generated jq change plus `scripts/PublicApiRemovedMarker.ps1` takes this full-validation path. The helper remains in the generator-source fingerprint and outside generated-output staging. Changes to shared generator inputs intentionally do not qualify for a single-integration-only build, so no allowlist exemption was added.

**Comment by thomhurst** [comment]
Pushed 69b264203b and started the required real Generate CLI Options validation for jq: https://github.com/thomhurst/ModularPipelines/actions/runs/34788997774. This run uses the repaired historical baselines and will open a generated-output PR if regeneration changes files. The outstanding review thread remains open pending the subsequent CI/review cycle.

**Comment by thomhurst** [comment]
Pushed 305f7d451598943be5a171dd95c381e7328b7c64 after diagnosing the new Fast fail failure in run 34792035048. GitHub's synthetic merge incorporated newer Podman baselines from main; 23 new removed entries consequently lacked their shipped declarations. I condensed the already-reviewed fix sequence, rebased its exact final tree onto current main, and reran Repair-PublicApiBaselineHistory.ps1 for Podman. The incremental repair adds 23 shipped declarations and changes no active APIs.

The repository validator now passes for all 61 package projects. An independent comparison against main confirms unchanged active APIs and all 23,469 unique removal records, with 23,290 restored shipped declarations. Four relevant PowerShell suites pass; Podman Release build has zero warnings/errors and all three integration tests pass.

The jq output from #4885 passed its full validation before incorporation, but this rebase changes generator inputs. Fresh authoritative jq generation is queued at https://github.com/thomhurst/ModularPipelines/actions/runs/34792972597 with automatic merging and coverage-shrinkage approval disabled. It must be incorporated and reviewed before merge; no provenance was hand-edited.

**Comment by thomhurst** [comment]
Addressed review 5657751611 in e2bc359e1b9099d0af89eb0d5ffa705e6f44a7e2, with one evidence-based correction to the requested whitespace normalization:

1. `Read-ActiveApi` must preserve raw API lines. The installed PublicApiAnalyzers 5.6.0 package identifies Roslyn commit c0573ed0a7dc3e3b4d2e70da47f97cc51a35524f in its nuspec. Its [ReadApiData implementation](https://github.com/dotnet/roslyn/blob/c0573ed0a7dc3e3b4d2e70da47f97cc51a35524f/src/RoslynAnalyzers/PublicApiAnalyzers/Core/Analyzers/DeclarePublicApiAnalyzer.cs) uses `line.ToString()`, skips whitespace-only lines, then tests the removal prefix and retains the remaining payload without trimming. An indented marker is not a removal, and a trailing space changes its payload. Trimming here would invent a retirement the analyzer does not recognize and break the existing exact-marker regression. I removed the inconsistent local trim instead and routed classification through the shared parser. New tests cover indented marker/header lines alongside the existing trailing-space case.
2. Blank/header/entry classification now lives in PublicApiRemovedMarker.ps1. The validator, merge script, history repair, and summary all consume it. Headers retain the existing column-zero convention; all readers now agree about indented lines. The new summary regression failed before this change and passes now.
3. Each negative child process now returns captured output and exit code, with the caller's prior LASTEXITCODE restored in a finally block at the invocation. The end-of-script reset is removed. A regression seeds caller status 23 and verifies it survives the negative check, preventing a later reset from hiding leakage.

The failed Ubuntu artifact from run 34795076606 identified Command_Header_Precedes_Streamed_Output_And_Completion as the sole failed core test. It now observes logger calls directly, keeps every header/output/completion ordering assertion, requires the command to succeed, and always disposes the pipeline. All 52 command logger tests pass. Six PowerShell suites and baseline validation for all 61 packages pass; the Release core/test solution builds with zero errors. Scoped formatting and diff checks pass with the default dotnet guard limits.

Fresh authoritative jq generation is running at https://github.com/thomhurst/ModularPipelines/actions/runs/34797964854 because these script changes affect generator provenance. That output and the new CI/review cycle remain pending before merge.
